File: 0739-daily-temperatures/0739-daily-temperatures.java
class Solution {
    public int[] dailyTemperatures(int[] temperatures) {
        int[] ans = new int[temperatures.length];
        Deque<Integer> stack = new ArrayDeque<>();
        
        for (int i = 0; i < temperatures.length; i++) {
            while (!stack.isEmpty() && temperatures[i] > temperatures[stack.peek()]) {
                ans[stack.peek()] = i - stack.peek();
                stack.pop();  
            }
            stack.push(i); // cannot use add, add is for Queue
        }
        return ans;
    }
}

File: 0739-daily-temperatures/0739-daily-temperatures.py
class Solution:
    def dailyTemperatures(self, temperatures: List[int]) -> List[int]:
        n = len(temperatures)
        ans = [0] * n
        monotonic_stack = deque()

        for i in range(n):
            while monotonic_stack and temperatures[i] > temperatures[monotonic_stack[-1]]:
                # 不能用这种写法，Python不像Java是按先后顺序的
                # ans[monotonic_stack[-1]] = i - monotonic_stack.pop()
                prev_day = monotonic_stack.pop()
                ans[prev_day] = i - prev_day
            monotonic_stack.append(i)

        return ans
